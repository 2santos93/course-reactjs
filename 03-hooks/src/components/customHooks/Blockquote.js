export const Blockquote = ({quote, author}) => {
    return (
        <blockquote className="blockquote">
            <p>{quote}</p>
            <footer className="blockquote-footer">{author}</footer>
        </blockquote>
    )
}
