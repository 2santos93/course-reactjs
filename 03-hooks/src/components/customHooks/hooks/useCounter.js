import {useState} from 'react';

export const useCounter = (initialCounter=1) => {
    const [counter, setCounter] = useState(initialCounter);

    const increment = (e) => {
        setCounter(cPrev => cPrev + 1);
    }

    return [counter, increment];
}
