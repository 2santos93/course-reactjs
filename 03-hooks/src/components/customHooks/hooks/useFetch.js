import {useState, useEffect, useRef} from 'react';

export const useFetch = (url) => {
    
    const isMounted = useRef(true);
    const [state, setState] = useState({loading: true, data: null, error:false});

    useEffect(() => {
        
        return () => {
            isMounted.current = false;
        }
    }, []);
    
    useEffect(() => {
        if(!isMounted) return;

        const getData = async () => {
            const response = await fetch(url);
            const data = await response.json();
            if(!isMounted) return;
            setState({loading: false, data, error:false});
        }
        setState({...state, loading:true})
        try{
            getData();
        }catch(e){
            if(!isMounted) return;
            setState({
                loading: false, 
                data: null,
                error: true
            })
        }


    }, [url]);

    return state;
}
