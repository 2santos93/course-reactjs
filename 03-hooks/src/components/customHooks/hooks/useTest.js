import { useEffect } from "react"

export const useTest = (load) => {

  useEffect(() => {
    console.log('me monte' +load);
    return () => {
      console.log('me desmonte'+ load);
    }
  }, [])

}
