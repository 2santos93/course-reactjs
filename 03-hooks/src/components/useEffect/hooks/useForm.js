import { useState } from "react"

export const useForm = (initialState={}) => {
    const [state, setState] = useState(initialState);
    
    // const onSubmitHandler = (e) => {
    //     e.preventDefault();
    //     console.log(e);
    // }

    const onChangeHandler = (e) => {
        const {name, value} = e.target;
        setState((prev) => ({
            ...prev,
            [name]: value
        }))
    }
    
    return [state, onChangeHandler];
}
