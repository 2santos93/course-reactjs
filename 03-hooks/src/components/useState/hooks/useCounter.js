import { useState } from "react";

const useCounter = (initialState = 10) => {

    const [state, setstate] = useState(initialState)

    const increment = (factor = 1) => {
        setstate(state + factor);
    }
    const decrement = (factor = 1) => {
        setstate(state - factor);
    }
    const reset = () => {
        setstate(initialState);
    }

    return {decrement, increment, reset, state};
}

export {useCounter};
