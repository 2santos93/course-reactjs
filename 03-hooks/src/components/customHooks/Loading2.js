import { useTest } from "./hooks/useTest"

export const Loading2 = () => {
    console.log('loading 2');

    useTest(2);

    return (
        <div className="spinner-border text-primary" role="status">
            <span className="visually-hidden">Loading...</span>
        </div>
    )
}
