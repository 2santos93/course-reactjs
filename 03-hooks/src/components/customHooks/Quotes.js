import {Loading} from './Loading';
import {Loading2} from './Loading2';
import { useCounter } from './hooks/useCounter';
import { useFetch } from './hooks/useFetch';
import { Blockquote } from './Blockquote';

export const Quotes = () => {
    const [counter, increment] = useCounter();

    const {loading, data} = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
    
    return (
        <>
            <h1>Breaking Bad</h1>
            <hr />

            {
                loading && <Loading />
            }

            {
                loading && <Loading2 />
            }

            {
                !!data && data.map( ({author, quote}) => (
                    <Blockquote key={Math.random()} author={author} quote={quote}/>
                ))
            }

            <button
                className="btn btn-primary"
                onClick={increment}
            >
                Next Quote
            </button>
        </>
    )
}
