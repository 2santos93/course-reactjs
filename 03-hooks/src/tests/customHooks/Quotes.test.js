import {shallow} from 'enzyme';
import { useCounter } from '../../components/customHooks/hooks/useCounter';
import { useFetch } from '../../components/customHooks/hooks/useFetch';
import { Quotes } from '../../components/customHooks/Quotes';

jest.mock('../../components/customHooks/hooks/useFetch');
jest.mock('../../components/customHooks/hooks/useCounter');

describe('Quotes test', () => {

  beforeEach( () => {
    jest.resetAllMocks();
    useCounter.mockReturnValue([1, () => {}]);
  })
  

  test('should render ok', () => {
    useFetch.mockReturnValue({
      loading: true,
      data: null,
      error: false
    })
    
    const wrapper = shallow(<Quotes />);
    
    expect(wrapper).toMatchSnapshot();
  })

  test('should render data', () => {
    useFetch.mockReturnValue({
      loading: false,
      data: [{author: 'nelson', quote: 'hello world'}],
      error: false
    })

    const wrapper = shallow(<Quotes />);

    expect(wrapper.find('Loading').exists()).toBe(false);
  })
  
})
