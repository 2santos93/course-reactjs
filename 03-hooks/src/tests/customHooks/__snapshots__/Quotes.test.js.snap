// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quotes test should render ok 1`] = `
<Fragment>
  <h1>
    Breaking Bad
  </h1>
  <hr />
  <Loading />
  <Loading2 />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next Quote
  </button>
</Fragment>
`;
